File: packages/hook/src/commands/exec.ts
/**
 * `chunk hook exec` command.
 *
 * Executes a shell command and enforces the result.
 *
 * Subcommands:
 *   - `run`       — Run command, save result, check result → fail on failure.
 *   - `run --no-check` — Run command, save result, skip the check. Always exits 0.
 *   - `check`     — Deferred check: read a saved result and fail on failure.
 *
 * Exit codes:
 *   0 — Pass / allow
 *   2 — Block / fail
 *   1 — Infra error
 */

import type { AgentEvent, HookAdapter } from "../lib/adapter";
import {
	blockNoCount,
	blockWithLimit,
	evaluateSentinel,
	guardStopEvent,
	matchesTrigger,
	resolveTriggerPatterns,
} from "../lib/check";
import type { ResolvedConfig, ResolvedExec } from "../lib/config";
import { getExec } from "../lib/config";
import { detectChanges, getChangedFiles, substitutePlaceholders } from "../lib/git";
import { log } from "../lib/log";
import { runCommand } from "../lib/proc";
import type { SentinelData } from "../lib/sentinel";
import { readSentinel, removeSentinel, resetBlockCount, writeSentinel } from "../lib/sentinel";
import { shellQuote } from "../lib/shell-env";
import { readMarker } from "./scope";

/** Build a name-qualified tag for log messages. */
function ntag(name: string): string {
	return `exec:${name}`;
}

/** CLI flags parsed from argv for the exec command. */
export type ExecFlags = {
	subcommand: "check" | "run";
	name: string;
	cmd?: string;
	timeout?: number;
	fileExt?: string;
	staged?: boolean;
	always?: boolean;
	noCheck?: boolean;
	on?: string;
	trigger?: string;
	limit?: number;
};

/**
 * Execute the exec command.
 */
export async function runExec(
	config: ResolvedConfig,
	adapter: HookAdapter,
	event: AgentEvent,
	flags: ExecFlags,
): Promise<void> {
	const t = ntag(flags.name);
	log(
		t,
		`subcommand=${flags.subcommand} event=${event.eventName || "(none)"} tool=${event.toolName || "(none)"}`,
	);

	// Resolve exec config (from YAML or CLI flags)
	const exec = resolveExecFromFlags(config, flags);

	switch (flags.subcommand) {
		case "check":
			return runCheck(config, adapter, event, flags, exec, t);
		case "run":
			if (flags.noCheck) {
				return runNoCheck(config, flags, exec, t);
			}
			return runFull(config, adapter, event, flags, exec, t);
	}
}

// ---------------------------------------------------------------------------
// Resolve exec config from flags + YAML
// ---------------------------------------------------------------------------

function resolveExecFromFlags(config: ResolvedConfig, flags: ExecFlags): ResolvedExec {
	const yamlExec = getExec(config, flags.name);
	return {
		command:
			flags.cmd ?? yamlExec?.command ?? `echo 'No command configured for exec: ${flags.name}'`,
		fileExt: flags.fileExt ?? yamlExec?.fileExt ?? "",
		always: flags.always ?? yamlExec?.always ?? false,
		timeout: flags.timeout ?? yamlExec?.timeout ?? 300,
		limit: flags.limit ?? yamlExec?.limit ?? 0,
	};
}

// ---------------------------------------------------------------------------
// check — read result file, emit decision (hook-facing)
// ---------------------------------------------------------------------------

async function runCheck(
	config: ResolvedConfig,
	adapter: HookAdapter,
	event: AgentEvent,
	flags: ExecFlags,
	exec: ResolvedExec,
	t: string,
): Promise<void> {
	const limit = flags.limit ?? exec.limit;
	guardStopEvent(t, adapter, event, limit);

	// Trigger matching
	const triggerPatterns = resolveTriggerPatterns(t, config, flags);
	if (triggerPatterns.length > 0 && !matchesTrigger(adapter, event, triggerPatterns)) {
		const cmd = adapter.commandSummary(event);
		log(t, `Trigger ${JSON.stringify(triggerPatterns)} did not match${cmd}, allowing`);
		adapter.allow();
	}

	// Skip-if-no-changes
	if (!exec.always) {
		const hasChanges = await detectChanges({
			cwd: config.projectDir,
			fileExt: exec.fileExt,
			staged: flags.staged,
		});
		if (!hasChanges) {
			log(t, "No changed files, allowing");
			adapter.allow();
		}
	}

	const sentinel = readSentinel(config.sentinelDir, config.projectDir, flags.name);

	// Session-aware staleness: sentinels from a different session are
	// treated as missing so the command re-runs with fresh context.
	const marker = readMarker(config.projectDir);
	const currentSessionId = marker?.sessionId;

	emitCheckResult(config, adapter, flags, exec, sentinel, t, currentSessionId);
}

/**
 * Evaluate a consumed sentinel and emit the check result (self-consuming).
 */
function emitCheckResult(
	config: ResolvedConfig,
	adapter: HookAdapter,
	flags: ExecFlags,
	exec: ResolvedExec,
	sentinel: SentinelData | undefined,
	t: string,
	currentSessionId?: string,
): never {
	const result = evaluateSentinel(sentinel, currentSessionId);
	const limit = flags.limit ?? exec.limit;
	const name = flags.name;

	switch (result.kind) {
		case "missing": {
			const runnerCmd = buildRunnerCommand(flags);
			const reason =
				`Exec "${name}" has no results. Run it first:\n\n` +
				`  ${runnerCmd}\n\n` +
				`Retry after the command completes.`;
			log(t, "Result: missing → action: block (agent must run command first)");
			blockNoCount(t, adapter, reason, config.projectDir);
			break;
		}
		case "pending": {
			const timeout = flags.timeout ?? exec.timeout;
			if (sentinel?.startedAt && timeout > 0) {
				const elapsed = (Date.now() - new Date(sentinel.startedAt).getTime()) / 1000;
				if (elapsed > timeout) {
					log(
						t,
						`Result: pending (timed out after ${Math.round(elapsed)}s, limit: ${timeout}s) → action: block`,
					);
					removeSentinel(config.sentinelDir, config.projectDir, name);
					const runnerCmd = buildRunnerCommand(flags);
					const reason =
						`Exec "${name}" timed out after ${Math.round(elapsed)}s ` +
						`(configured timeout: ${timeout}s).\n\n` +
						`The previous run may have an issue (infinite loop, deadlock, etc.). ` +
						`Investigate and re-run:\n\n` +
						`  ${runnerCmd}`;
					blockWithLimit(t, adapter, config, name, limit, reason);
					break;
				}
			}
			log(t, "Result: pending → action: block (waiting for command to complete)");
			blockNoCount(
				t,
				adapter,
				`Exec "${name}" is still running. Wait for completion before retrying.`,
				config.projectDir,
			);
			break;
		}
		case "pass":
			log(t, "Result: pass → action: allow");
			resetBlockCount(config.sentinelDir, config.projectDir, name);
			removeSentinel(config.sentinelDir, config.projectDir, name);
			adapter.allow();
			break;
		case "fail": {
			const reason = formatFailureReason(
				name,
				result.sentinel.command ?? exec.command,
				result.sentinel.exitCode ?? 1,
				result.sentinel.output ?? "(no output)",
			);
			log(
				t,
				`Result: fail (exit ${result.sentinel.exitCode}) → action: block (agent must fix and re-run)`,
			);
			blockWithLimit(t, adapter, config, name, limit, reason);
			break;
		}
	}
}

/**
 * Evaluate a sentinel and emit the result (direct, no gate).
 * Used by `runFull` after executing a command.
 */
function emitSentinelResult(
	config: ResolvedConfig,
	adapter: HookAdapter,
	flags: ExecFlags,
	exec: ResolvedExec,
	sentinel: SentinelData | undefined,
	t: string,
): never {
	const result = evaluateSentinel(sentinel);
	const limit = flags.limit ?? exec.limit;
	const name = flags.name;

	switch (result.kind) {
		case "missing": {
			const runnerCmd = buildRunnerCommand(flags);
			const reason =
				`Exec "${name}" has no results. Run it first:\n\n` +
				`  ${runnerCmd}\n\n` +
				`Retry after the command completes.`;
			log(t, "Result: missing (unexpected in direct path) → action: block");
			blockNoCount(t, adapter, reason, config.projectDir);
			break;
		}
		case "pending":
			log(t, "Result: pending (unexpected in direct path) → action: block");
			blockNoCount(
				t,
				adapter,
				`Exec "${name}" is still running. Wait for completion before retrying.`,
				config.projectDir,
			);
			break;
		case "pass":
			log(t, "Result: pass → action: allow");
			resetBlockCount(config.sentinelDir, config.projectDir, name);
			adapter.allow();
			break;
		case "fail": {
			const reason = formatFailureReason(
				name,
				result.sentinel.command ?? exec.command,
				result.sentinel.exitCode ?? 1,
				result.sentinel.output ?? "(no output)",
			);
			log(
				t,
				`Result: fail (exit ${result.sentinel.exitCode}) → action: block (agent must fix and re-run)`,
			);
			blockWithLimit(t, adapter, config, name, limit, reason);
			break;
		}
	}
}

// ---------------------------------------------------------------------------
// run --no-check
// ---------------------------------------------------------------------------

async function runNoCheck(
	config: ResolvedConfig,
	flags: ExecFlags,
	exec: ResolvedExec,
	t: string,
): Promise<void> {
	const { sentinel } = await executeExec(config, flags, exec, t);

	if (sentinel.skipped) {
		log(t, "No changed files, skipped");
	} else if (sentinel.status === "pass") {
		log(t, "Result: pass (no-check mode)");
	} else {
		log(t, `Result: fail (exit ${sentinel.exitCode}, no-check mode)`);
	}
	process.exit(0);
}

// ---------------------------------------------------------------------------
// run (default) — run command + write result + output decision
// ---------------------------------------------------------------------------

async function runFull(
	config: ResolvedConfig,
	adapter: HookAdapter,
	event: AgentEvent,
	flags: ExecFlags,
	exec: ResolvedExec,
	t: string,
): Promise<void> {
	const limit = flags.limit ?? exec.limit;
	guardStopEvent(t, adapter, event, limit);

	// Trigger matching (align with check semantics)
	const triggerPatterns = resolveTriggerPatterns(t, config, flags);
	if (triggerPatterns.length > 0 && !matchesTrigger(adapter, event, triggerPatterns)) {
		const cmd = adapter.commandSummary(event);
		log(t, `Trigger ${JSON.stringify(triggerPatterns)} did not match${cmd}, allowing`);
		adapter.allow();
	}

	const { sentinel } = await executeExec(config, flags, exec, t);
	emitSentinelResult(config, adapter, flags, exec, sentinel, t);
}

// ---------------------------------------------------------------------------
// Shared exec execution core
// ---------------------------------------------------------------------------

type ExecResult = {
	sentinel: SentinelData;
};

async function executeExec(
	config: ResolvedConfig,
	flags: ExecFlags,
	exec: ResolvedExec,
	t: string,
): Promise<ExecResult> {
	const startedAt = new Date().toISOString();

	// Read the current session ID from the scope marker so it can be
	// stored in sentinels for session-aware staleness detection.
	const sessionId = readMarker(config.projectDir)?.sessionId;

	// Skip-if-no-changes — return a synthetic sentinel without writing it to disk.
	// Writing a "pass" sentinel here would be dangerous: an agent could run the
	// command while the repo is clean to farm a passing sentinel, then introduce
	// bugs and commit — the stale "pass" sentinel would allow the commit.
	// The check and sync paths have their own independent `detectChanges`
	// short-circuits, so they never read a skipped sentinel.
	if (!exec.always) {
		const hasChanges = await detectChanges({
			cwd: config.projectDir,
			fileExt: exec.fileExt,
			staged: flags.staged,
		});
		if (!hasChanges) {
			const sentinel: SentinelData = {
				status: "pass",
				startedAt,
				finishedAt: new Date().toISOString(),
				exitCode: 0,
				output: "No changed files. Skipped.",
				skipped: true,
				project: config.projectDir,
				sessionId,
			};
			return { sentinel };
		}
	}

	// Write pending result
	writeSentinel(config.sentinelDir, config.projectDir, flags.name, {
		status: "pending",
		startedAt,
		project: config.projectDir,
		sessionId,
	});

	const command = await buildCommand(config, exec, flags);
	log(t, `Running: ${command}`);

	const runStart = Date.now();
	const result = await runCommand({
		command,
		cwd: config.projectDir,
		timeout: exec.timeout,
	});
	const elapsedMs = Date.now() - runStart;

	const status = result.exitCode === 0 ? "pass" : "fail";
	const sentinel: SentinelData = {
		status,
		startedAt,
		finishedAt: new Date().toISOString(),
		exitCode: result.exitCode,
		command: result.command,
		output: result.output,
		project: config.projectDir,
		sessionId,
	};
	writeSentinel(config.sentinelDir, config.projectDir, flags.name, sentinel);
	log(t, `Completed: ${status} (exit ${result.exitCode}, ${elapsedMs}ms)`);

	return { sentinel };
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/** Build the shell command with placeholder substitution. */
async function buildCommand(
	config: ResolvedConfig,
	exec: ResolvedExec,
	flags: ExecFlags,
): Promise<string> {
	let command = flags.cmd ?? exec.command;

	if (command.includes("{{CHANGED_FILES}}") || command.includes("{{CHANGED_PACKAGES}}")) {
		const files = await getChangedFiles({
			cwd: config.projectDir,
			stagedOnly: flags.staged ?? false,
			fileExt: exec.fileExt,
		});
		command = substitutePlaceholders(command, files);
	}

	return command;
}

/**
 * Build the `run --no-check` command string for the "missing" block message.
 */
export function buildRunnerCommand(flags: ExecFlags): string {
	const parts = ["chunk hook exec run", flags.name, "--no-check"];
	if (flags.cmd) parts.push(`--cmd '${shellQuote(flags.cmd)}'`);
	if (flags.timeout !== undefined) parts.push(`--timeout ${flags.timeout}`);
	if (flags.fileExt) parts.push(`--file-ext '${shellQuote(flags.fileExt)}'`);
	if (flags.staged) parts.push("--staged");
	if (flags.always) parts.push("--always");
	return parts.join(" ");
}

/** Format a concise failure reason for the agent. */
export function formatFailureReason(
	name: string,
	command: string,
	exitCode: number,
	output: string,
): string {
	const header =
		exitCode === 124
			? `Exec "${name}" timed out (command: ${command}).`
			: `Exec "${name}" failed (exit ${exitCode}, command: ${command}).`;
	return `${header} Fix the issues and retry.\n\nOutput:\n${output}`;
}

File: internal/hook/exec.go
package hook

import (
	"context"
	"errors"
	"fmt"
	"log/slog"
	"math"
	"os"
	"os/exec"
	"strings"
	"time"
)

// ExecRunFlags holds parsed flags for exec run.
type ExecRunFlags struct {
	Name    string
	Cmd     string
	Timeout int
	FileExt string
	Staged  bool
	Always  bool
	NoCheck bool
	On      string
	Trigger string
	Limit   int
	Matcher string
}

// ExecCheckFlags holds parsed flags for exec check.
type ExecCheckFlags struct {
	Name         string
	Timeout      int
	FileExt      string
	Staged       bool
	Always       bool
	On           string
	Trigger      string
	Limit        int
	Matcher      string
	Cmd          string
	AllowMissing bool
}

// resolveExec merges flags with config to produce effective exec settings.
func resolveExec(cfg *ResolvedConfig, name, flagCmd, flagFileExt string, flagAlways bool, flagTimeout, flagLimit int) ExecConfig {
	yamlExec, ok := cfg.Execs[name]
	if !ok {
		yamlExec = ExecConfig{Timeout: 300}
	}

	command := flagCmd
	if command == "" {
		command = yamlExec.Command
	}
	if command == "" {
		command = fmt.Sprintf("echo 'No command configured for exec: %s'", name)
	}

	fileExt := flagFileExt
	if fileExt == "" {
		fileExt = yamlExec.FileExt
	}

	always := flagAlways || yamlExec.Always

	timeout := flagTimeout
	if timeout == 0 {
		timeout = yamlExec.Timeout
	}
	if timeout == 0 {
		timeout = 300
	}

	limit := flagLimit
	if limit == 0 {
		limit = yamlExec.Limit
	}

	return ExecConfig{
		Command: command,
		FileExt: fileExt,
		Always:  always,
		Timeout: timeout,
		Limit:   limit,
	}
}

// executeAndSave runs a command, writes the sentinel, and returns the verdict.
func executeAndSave(cfg *ResolvedConfig, execCfg ExecConfig, name string, staged bool) ExecCheckVerdict {
	startedAt := time.Now().UTC().Format(time.RFC3339)

	marker := ReadMarker(cfg.ProjectDir)
	sessionID := ""
	if marker != nil {
		sessionID = marker.SessionID
	}

	// Write pending sentinel
	if err := writeSentinel(cfg.SentinelDir, cfg.ProjectDir, name, SentinelData{
		Status:    "pending",
		StartedAt: startedAt,
		Project:   cfg.ProjectDir,
		SessionID: sessionID,
	}); err != nil {
		fmt.Fprintf(os.Stderr, "chunk: warning: could not write sentinel for %q: %v\n", name, err)
	}

	// Build command with placeholder substitution
	command := execCfg.Command
	if strings.Contains(command, "{{CHANGED_FILES}}") || strings.Contains(command, "{{CHANGED_PACKAGES}}") {
		files := getChangedFiles(cfg.ProjectDir, staged, execCfg.FileExt)
		command = substitutePlaceholders(command, files)
	}

	// Execute command with timeout
	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(execCfg.Timeout)*time.Second)
	defer cancel()

	cmd := exec.CommandContext(ctx, "sh", "-c", command)
	cmd.Dir = cfg.ProjectDir
	cmd.Env = cleanEnv()
	output, err := cmd.CombinedOutput()

	exitCode := 0
	if err != nil {
		var exitErr *exec.ExitError
		if errors.As(err, &exitErr) {
			exitCode = exitErr.ExitCode()
		} else {
			exitCode = 1
		}
	}

	status := verdictPass
	if exitCode != 0 {
		status = verdictFail
	}

	contentHash := computeFingerprint(cfg.ProjectDir, staged, execCfg.FileExt)

	sentinel := SentinelData{
		Status:            status,
		StartedAt:         startedAt,
		FinishedAt:        time.Now().UTC().Format(time.RFC3339),
		ExitCode:          exitCode,
		Command:           command,
		ConfiguredCommand: execCfg.Command,
		Output:            strings.TrimRight(string(output), "\n"),
		Project:           cfg.ProjectDir,
		SessionID:         sessionID,
		ContentHash:       contentHash,
	}
	if err := writeSentinel(cfg.SentinelDir, cfg.ProjectDir, name, sentinel); err != nil {
		fmt.Fprintf(os.Stderr, "chunk: warning: could not write sentinel for %q: %v\n", name, err)
	}

	return ExecCheckVerdict{Kind: status, Sentinel: &sentinel}
}

// RunExecRun executes a configured command, saves result as sentinel.
func RunExecRun(cfg *ResolvedConfig, flags ExecRunFlags) error {
	if !IsEnabled(flags.Name) {
		slog.Info("hook skipped", "name", flags.Name, "reason", "not enabled")
		return nil
	}

	execCfg := resolveExec(cfg, flags.Name, flags.Cmd, flags.FileExt, flags.Always, flags.Timeout, flags.Limit)

	// Skip if no changes (unless always)
	if !execCfg.Always {
		hasChanges, _ := detectChanges(cfg.ProjectDir, execCfg.FileExt, flags.Staged)
		if !hasChanges {
			startedAt := time.Now().UTC().Format(time.RFC3339)
			marker := ReadMarker(cfg.ProjectDir)
			sessionID := ""
			if marker != nil {
				sessionID = marker.SessionID
			}
			contentHash := computeFingerprint(cfg.ProjectDir, flags.Staged, execCfg.FileExt)
			if err := writeSentinel(cfg.SentinelDir, cfg.ProjectDir, flags.Name, SentinelData{
				Status:      "pass",
				StartedAt:   startedAt,
				FinishedAt:  time.Now().UTC().Format(time.RFC3339),
				ExitCode:    0,
				Output:      "No changed files. Skipped.",
				Skipped:     true,
				Project:     cfg.ProjectDir,
				SessionID:   sessionID,
				ContentHash: contentHash,
			}); err != nil {
				fmt.Fprintf(os.Stderr, "chunk: warning: could not write sentinel for %q: %v\n", flags.Name, err)
			}
			return nil
		}
	}

	verdict := executeAndSave(cfg, execCfg, flags.Name, flags.Staged)

	if flags.NoCheck {
		return nil
	}

	if verdict.Kind == verdictFail {
		exitCode := 1
		if verdict.Sentinel != nil {
			exitCode = verdict.Sentinel.ExitCode
		}
		return fmt.Errorf("exec %q failed (exit %d)", flags.Name, exitCode)
	}
	return nil
}

// RunExecCheck reads a saved sentinel and enforces the result.
func RunExecCheck(cfg *ResolvedConfig, flags ExecCheckFlags, event map[string]interface{}) error {
	if !IsEnabled(flags.Name) {
		slog.Info("hook skipped", "name", flags.Name, "reason", "not enabled")
		return nil
	}

	execCfg := resolveExec(cfg, flags.Name, flags.Cmd, flags.FileExt, flags.Always, flags.Timeout, flags.Limit)
	limit := execCfg.Limit

	// Stop-event recursion guard
	if resp := guardStopEvent(event, limit); resp != nil {
		return emitResponse(*resp)
	}

	verdict := preEvaluateExec(cfg, event, execCfg, flags.Name, flags.Staged, flags.On, flags.Trigger, nil, "")

	return emitExecVerdict(cfg, flags, execCfg, verdict)
}

func emitExecVerdict(cfg *ResolvedConfig, flags ExecCheckFlags, execCfg ExecConfig, verdict ExecCheckVerdict) error {
	limit := execCfg.Limit
	name := flags.Name

	switch verdict.Kind {
	case "skip-trigger":
		return nil // allow
	case "skip-no-changes":
		return nil // allow
	case verdictMissing:
		if flags.AllowMissing {
			return nil // nothing was run, nothing to enforce
		}
		verdict = executeAndSave(cfg, execCfg, name, flags.Staged)
		return emitExecVerdict(cfg, flags, execCfg, verdict)
	case "pending":
		stillRunning := fmt.Sprintf("Exec %q is still running. Wait for completion before retrying.", name)
		sentinel := verdict.Sentinel
		timeout := execCfg.Timeout
		if sentinel == nil || sentinel.StartedAt == "" || timeout <= 0 {
			return emitResponse(blockNoCount(cfg.ProjectDir, stillRunning))
		}
		started, err := time.Parse(time.RFC3339, sentinel.StartedAt)
		if err != nil {
			return emitResponse(blockNoCount(cfg.ProjectDir, stillRunning))
		}
		elapsed := time.Since(started).Seconds()
		if elapsed <= float64(timeout) {
			return emitResponse(blockNoCount(cfg.ProjectDir, stillRunning))
		}
		runnerCmd := buildRunnerCommand(name, flags.Cmd, flags.Timeout, flags.FileExt, flags.Staged, flags.Always)
		reason := fmt.Sprintf(
			"Exec %q timed out after %ds (configured timeout: %ds).\n\n"+
				"The previous run may have an issue (infinite loop, deadlock, etc.). "+
				"Investigate and re-run:\n\n  %s",
			name, int(math.Round(elapsed)), timeout, runnerCmd)
		return emitResponse(blockWithLimit(cfg, name, limit, reason))
	case verdictPass:
		resetBlockCount(cfg.SentinelDir, cfg.ProjectDir, name)
		return nil // allow
	case verdictFail:
		sentinel := verdict.Sentinel
		cmd := execCfg.Command
		exitCode := 1
		output := "(no output)"
		if sentinel != nil && sentinel.Command != "" {
			cmd = sentinel.Command
		}
		if sentinel != nil && sentinel.ExitCode != 0 {
			exitCode = sentinel.ExitCode
		}
		if sentinel != nil && sentinel.Output != "" {
			output = sentinel.Output
		}
		reason := formatFailureReason(name, cmd, exitCode, output)
		return emitResponse(blockWithLimit(cfg, name, limit, reason))
	}

	return nil
}

func buildRunnerCommand(name, cmd string, _ int, _ string, staged, always bool) string {
	parts := []string{"chunk validate", name, "--no-check"}
	if cmd != "" {
		parts = append(parts, fmt.Sprintf("--cmd '%s'", cmd))
	}
	if staged {
		parts = append(parts, "--staged")
	}
	if always {
		parts = append(parts, "--always")
	}
	return strings.Join(parts, " ")
}

func formatFailureReason(name, command string, exitCode int, output string) string {
	header := fmt.Sprintf("Exec %q failed (exit %d, command: %s).", name, exitCode, command)
	if exitCode == 124 {
		header = fmt.Sprintf("Exec %q timed out (command: %s).", name, command)
	}
	return fmt.Sprintf("%s Fix the issues and retry.\n\nOutput:\n%s", header, output)
}

// cleanEnv returns a filtered copy of the current environment,
// removing CHUNK_HOOK_* vars to prevent contamination.
func cleanEnv() []string {
	var env []string
	for _, e := range os.Environ() {
		if strings.HasPrefix(e, "CHUNK_HOOK_") {
			continue
		}
		env = append(env, e)
	}
	return env
}
